hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java

```diff
@@ -0,0 +1,640 @@
+    if (nsSummary == null) {
+      return 0;
+    }
+    Set<Long> subdirs = nsSummary.getChildDir();
+    int totalCnt = subdirs.size();
+    for (long subdir: subdirs) {
+      totalCnt += getTotalDirCount(subdir);
+    }
+    return totalCnt;
+  }
+
+  /**
+   * Given an object ID, return total data size (no replication)
+   * under this object.
+   * @param objectId the object's ID
+   * @return total used data size in bytes
+   * @throws IOException ioEx
+   */
+  private long getTotalSize(long objectId) throws IOException {
+    NSSummary nsSummary = reconNamespaceSummaryManager.getNSSummary(objectId);
+    if (nsSummary == null) {
+      return 0L;
+    }
+    long totalSize = nsSummary.getSizeOfFiles();
+    for (long childId: nsSummary.getChildDir()) {
+      totalSize += getTotalSize(childId);
+    }
+    return totalSize;
+  }
+
+  /**
+   * Given an object ID, return the file size distribution.
+   * @param objectId the object's ID
+   * @return int array indicating file size distribution
+   * @throws IOException ioEx
+   */
+  private int[] getTotalFileSizeDist(long objectId) throws IOException {
+    NSSummary nsSummary = reconNamespaceSummaryManager.getNSSummary(objectId);
+    if (nsSummary == null) {
+      return new int[ReconConstants.NUM_OF_BINS];
+    }
+    int[] res = nsSummary.getFileSizeBucket();
+    for (long childId: nsSummary.getChildDir()) {
+      int[] subDirFileSizeDist = getTotalFileSizeDist(childId);
+      for (int i = 0; i < ReconConstants.NUM_OF_BINS; ++i) {
+        res[i] += subDirFileSizeDist[i];
+      }
+    }
+    return res;
+  }
+
+  private List<OmBucketInfo> listBucketsUnderVolume(final String volumeName)
+          throws IOException {
+    List<OmBucketInfo> result = new ArrayList<>();
+    if (Strings.isNullOrEmpty(volumeName)
+            || !checkVolumeExistence(volumeName)) {
+      return null;
+    }
+    Table bucketTable = omMetadataManager.getBucketTable();
+    String seekPrefix =
+            omMetadataManager.getVolumeKey(volumeName + OM_KEY_PREFIX);
+
+    TableIterator<String, ? extends Table.KeyValue<String, OmBucketInfo>>
+            iterator = bucketTable.iterator();
+
+
+    while (iterator.hasNext()) {
+      Table.KeyValue<String, OmBucketInfo> kv = iterator.next();
+
+      String key = kv.getKey();
+      OmBucketInfo omBucketInfo = kv.getValue();
+
+      if (omBucketInfo != null) {
+        // We should return only the keys, whose keys match with the seek prefix
+        if (key.startsWith(seekPrefix)) {
+          result.add(omBucketInfo);
+        }
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Helper function to check if a path is a directory, key, or invalid.
+   * @param volName volume name
+   * @param bucketName bucket name
+   * @param keyName key name
+   * @return DIRECTORY, KEY, or UNKNOWN
+   * @throws IOException
+   */
+  private EntityType determineKeyPath(String volName, String bucketName,
```

Comment:
Can we use `OMFileRequest#getOMKeyInfoIfExists` here directly?

Only `volumeTable` and `bucketTable` are [`FULL_CACHE`](https://github.com/yuangu002/ozone/blob/afce21398041c85f3ffe5753336a65b317cf0756/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L444-L455).

`dirTable` and `fileTable` are [`PARTIAL_CACHE`] (by default), so we can just use `.get()`.  to avoid code duplication.

Reply:
> Can we use `OMFileRequest#getOMKeyInfoIfExists` inside?
> 
> Only `volumeTable` and `bucketTable` are [`FULL_CACHE`](https://github.com/yuangu002/ozone/blob/afce21398041c85f3ffe5753336a65b317cf0756/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmMetadataManagerImpl.java#L444-L455).
> 
> `dirTable` and `fileTable` are `PARTIAL_CACHE` (by default), so we can just use `.get()`. Deduplicate the function.

Reasons that I didn't do that:
1. `OMFileRequest#getOMKeyInfoIfExists` contains a `validateBucket` method, which throws OM exception if bucket doesn't exist (in this case, the bucket won't exist, because `validateBucket` uses `get()` to check bucket existence). Plus, we also don't want to throw any `OMException` on the Recon side. Instead, we only want to set `EntityType` as `Unknown`.
2. There's no need for us to instantiate `OzoneFileStatus` class as `getOMKeyInfoIfExists` did. The only thing we want to do is just to distinguish directory from key.

Reply:
interesting. this leaves us with some refactoring work to do with OMFileRequest then.

Ideally we want `determineKeyPath` and `getOMKeyInfoIfExists` calling a common helper function (in OMFileRequest) to walk the FS tree. otherwise, a future bug fix/improvement in `getOMKeyInfoIfExists` will almost certainly be **missed** by Recon.